lightning-invoice/src/utils.rs

```diff
@@ -255,4 +365,135 @@ mod test {
+		for channel in channels_2.iter() {
+			route_hints.push((nodes[2].node.get_phantom_scid(), nodes[2].node.get_our_node_id(), &channel));
+		}
+		let invoice = ::utils::create_phantom_invoice::<EnforcingSigner, &test_utils::TestKeysInterface>(Some(payment_amt), "test".to_string(), payment_hash, payment_secret, &route_hints, &nodes[1].keys_manager, Currency::BitcoinTestnet).unwrap();
+
+		assert_eq!(invoice.min_final_cltv_expiry(), MIN_FINAL_CLTV_EXPIRY as u64);
+		assert_eq!(invoice.description(), InvoiceDescription::Direct(&Description("test".to_string())));
+		assert_eq!(invoice.route_hints().len(), 2);
+		assert!(!invoice.features().unwrap().supports_basic_mpp());
+
+		let payee = Payee::from_node_id(invoice.recover_payee_pub_key())
+			.with_features(invoice.features().unwrap().clone())
+			.with_route_hints(invoice.route_hints());
+		let params = RouteParameters {
+			payee,
+			final_value_msat: invoice.amount_milli_satoshis().unwrap(),
+			final_cltv_expiry_delta: invoice.min_final_cltv_expiry() as u32,
+		};
+		let first_hops = nodes[0].node.list_usable_channels();
+		let network_graph = node_cfgs[0].network_graph;
+		let logger = test_utils::TestLogger::new();
+		let scorer = test_utils::TestScorer::with_fixed_penalty(0);
+		let route = find_route(
+			&nodes[0].node.get_our_node_id(), &params, network_graph,
+			Some(&first_hops.iter().collect::<Vec<_>>()), &logger, &scorer,
+		).unwrap();
+		let (payment_event, fwd_idx) = {
+			let mut payment_hash = PaymentHash([0; 32]);
+			payment_hash.0.copy_from_slice(&invoice.payment_hash().as_ref()[0..32]);
+			nodes[0].node.send_payment(&route, payment_hash, &Some(invoice.payment_secret().clone())).unwrap();
+			let mut added_monitors = nodes[0].chain_monitor.added_monitors.lock().unwrap();
+			assert_eq!(added_monitors.len(), 1);
+			added_monitors.clear();
+
+			let mut events = nodes[0].node.get_and_clear_pending_msg_events();
+			assert_eq!(events.len(), 1);
+			let fwd_idx = match events[0] {
+				MessageSendEvent::UpdateHTLCs { node_id, .. } => {
+					if node_id == nodes[1].node.get_our_node_id() {
+						1
+					} else { 2 }
+				},
+				_ => panic!("Unexpected event")
+			};
+			(SendEvent::from_event(events.remove(0)), fwd_idx)
+		};
+		nodes[fwd_idx].node.handle_update_add_htlc(&nodes[0].node.get_our_node_id(), &payment_event.msgs[0]);
+		commitment_signed_dance!(nodes[fwd_idx], nodes[0], &payment_event.commitment_msg, false, true);
+		expect_pending_htlcs_forwardable!(nodes[fwd_idx]);
+		let payment_preimage_opt = if user_generated_pmt_hash { None } else { Some(payment_preimage) };
+		let events = nodes[fwd_idx].node.get_and_clear_pending_events();
+		assert_eq!(events.len(), 2);
+		match events[0] {
+			Event::PendingHTLCsForwardable { .. } => { },
+			_ => panic!("Unexpected event"),
+		}
+		match events[1] {
+			Event::PaymentReceived { payment_hash: ref hash, ref purpose, amt } => {
```

Comment:
The way this is set up, does that work too for multi-part payments to a phantom invoice where the parts come in via distinct real nodes?

Reply:
No. In theory we could adapt it to work with users doing the work to claim across nodes, but I'm not sure what the use-case there is? All the nice "node offline -> seamless fallback to other nodes" stuff suddenly doesn't work. You still get shared liquidity, but that feels like it could just as easily be done with a single node at that point, no?

Reply:
Senders are free to choose routes right? So they may decide to use a different final hop towards the phantom node for some of the parts. I don't think you can force them not to, and they also won't be aware that they shouldn't.

One thing that I can think of is when node A doesn't have enough (cheap) liquidity and the sender decides to split the payment to also use up some of node B's liquidity.

Reply:
An alternative design is to make the phantom node a stand-alone process that communicates with each of the gateway nodes and is backed by a replicated database?

Reply:
No, you can't force them not to, but you can disable MPP, which accomplishes the same :)

Reply:
The sender may be short on liquidity towards their peers and might only be able to complete the payment via MPP.

If you run a business that is big enough to require a multi-node setup, I'd think it is undesirable to force all customers to pay in single shots only?